### PR TITLE
feat: Software Data - Rebased

### PR DESCRIPTION
**Pull request by RickiJay-WMDE**
* Move software to its own table; pull metadata from MediaWiki for extensions
* Return that metadata in an extension list
* Test all of that

For the record, about 11k lines of this are test data files

### CODE REVIEW

File: model/database/wikibase_software/software_tag_model.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+"""Wikibase Software Version Table"""
```
Comment (jonamar) [inline model/database/wikibase_software/software_tag_model.py:1]: should be software tag table not version table. :)

**Review by jonamar — DISMISSED** [review]

File: model/database/wikibase_software/software_tag_model.py (hunk 1 above)
Comment (RickiJay-WMDE) [reply model/database/wikibase_software/software_tag_model.py]: Updated!

**Review by tarrow — COMMENTED** [review]
I've read through the code and think I have a vague idea of what's going on. It's been a while since I last wrangled much python so I've not made any judgement calls about how pythonic or not the code is.

Allow me to write what I think this PR is about; if I'm correct maybe some of it could make it into either some `md` or some of the doc strings?

This code is designed to take the extensions (and skins? I'm not sure? Also maybe php libraries?) that are discovered on Wikibases you find in the wild. It then goes to mediawiki.org (and only mediawiki.org ?) to look at the extension metadata page. It scrapes and parses that page in an attempt to create a local, structured, copy of some of the data in it (e.g latest version, download count etc.).

Naturally none of this is blocking but maybe it gives you a feeling for how this appears to a different pair of eyes. I wonder if it would be clearer to the reader if there was a harder separation between scraping Wikibases for information from them and looking up stuff on mediawiki.org about their configuration?

It did make me think that having this data you're scraping from extension pages would make sense to keep in a public, structured form. I wonder if it would actually happily live on either Wikidata or on wikibase.world.

**Review by tarrow — COMMENTED** [review]
I've read through the code and think I have a vague idea of what's going on. It's been a while since I last wrangled much python so I've not made any judgement calls about how pythonic or not the code is.

Allow me to write what I think this PR is about; if I'm correct maybe some of it could make it into either some `md` or some of the doc strings?

This code is designed to take the extensions (and skins? I'm not sure? Also maybe php libraries?) that are discovered on Wikibases you find in the wild. It then goes to mediawiki.org (and only mediawiki.org ?) to look at the extension metadata page. It scrapes and parses that page in an attempt to create a local, structured, copy of some of the data in it (e.g latest version, download count etc.).

Naturally none of this is blocking but maybe it gives you a feeling for how this appears to a different pair of eyes. I wonder if it would be clearer to the reader if there was a harder separation between scraping Wikibases for information from them and looking up stuff on mediawiki.org about their configuration?

It did make me think that having this data you're scraping from extension pages would make sense to keep in a public, structured form. I wonder if it would actually happily live on either Wikidata or on wikibase.world.

File: alembic/versions/7fd538ca8735_no_name_or_type.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        batch_op.create_unique_constraint(
+            "unique_observation_software_type_name",
+            [
+                "wikibase_software_version_observation_id",
+                "software_type",
+                "software_name",
+            ],
+        )
+
+    # update_query = sa.update(WikibaseSoftwareVersionModel).values(
```
[the remark this answers is not in the document]
Comment (RickiJay-WMDE) [reply alembic/versions/7fd538ca8735_no_name_or_type.py:54]: If we ever revert the code back to this commit, it'll be useful - I'd rather use the structured query than the text one below

File: data/gx/validation_definitions/wikibase_software_validation_definition.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{
```
[the remark this answers is not in the document]
Comment (RickiJay-WMDE) [reply data/gx/validation_definitions/wikibase_software_validation_definition.json:1]: They're manually generated, but not exactly hand-written. Just checking that the database matches what we expect.

File: fetch_data/soup_data/software/create_software_version_data_observation.py [hunk 4]
```diff
@@ -39,16 +49,18 @@ async def create_software_version_observation(wikibase_id: int) -> bool:
 
             observation.returned_data = True
 
-            installed_software_versions = compile_installed_software_versions(soup)
+            installed_software_versions = await compile_installed_software_versions(
```
[the remark this answers is not in the document]
Comment (RickiJay-WMDE) [reply fetch_data/soup_data/software/create_software_version_data_observation.py:52]: Yeah, it's this section of a Special:Version page:
image

I'm not sure of a better term for them - they're just called "software" in the page, but that really doesn't distinguish them from skins, libraries, and extensions. Any suggestions?

**Comment by RickiJay-WMDE** [comment]
> Allow me to write what I think this PR is about; if I'm correct maybe some of it could make it into either some `md` or some of the doc strings?

Thanks for reminding me to update the documentation! I had completely forgotten. :sweat_smile:

File: fetch_data/utils/clean_string.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+"""Clean String"""
+
+import re
+
+
+def clean_string(input_str: str) -> str:
+    """Clean String"""
```
Comment (rti) [inline fetch_data/utils/clean_string.py:7]:
```suggestion
   """
   Cleans the input string by normalizing whitespace and adjusting capitalization.
   - Collapses multiple spaces and replaces non-breaking spaces with single spaces.
   - Trims leading and trailing spaces.
   - Converts the string to uppercase if its length is less than 2.
   - Ensures the first character is capitalized while keeping the rest unchanged.
   """```

suggestion of a more descriptive comment

File: fetch_data/soup_data/software/create_software_version_data_observation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rti) [reply fetch_data/soup_data/software/create_software_version_data_observation.py:54]: I understand the confusion. I like the comment you just added above. I think now it is clear enough.

[the remark this answers is not in the document]
Comment (RickiJay-WMDE) [reply fetch_data/soup_data/software/create_software_version_data_observation.py:54]: Added a comment to clarify

**Comment by rti** [comment]
The `wikibase_software` table in `wikibase-test-data.db` contains two rows. Is this intended?

image

**Comment by rti** [comment]
Running the following query in the provided `wikibase-data.db`

```graphql
query {
  wikibase(wikibaseId: 1) {
    urls {
      baseUrl
    }
    softwareVersionObservations {
      mostRecent {
        installedExtensions {
          software {
            softwareName
          }
        }
      }
    }
  }
}
```
returns 
```json
{
  "data": {
    "wikibase": {
      "urls": {
        "baseUrl": "https://portal.mardi4nfdi.de"
      },
      "softwareVersionObservations": {
        "mostRecent": {
          "installedExtensions": [
            {
              "software": {
                "softwareName": "Echo"
              }
            },
            {
              "software": {
                "softwareName": "MathSearch"
              }
            },
            {
              "software": {
                "softwareName": "MatomoAnalytics"
              }
            },
            {
              "software": {
                "softwareName": "Nuke"
              }
            }
          ]
        }
      }
    }
  }
}
```
So it reports only 4 extensions. Is it because the site uses a different skin?

image

**Comment by RickiJay-WMDE** [comment]
> So it reports only 4 extensions. Is it because the site uses a different skin?

I think it's because I only check the first extension table. Whoops. Fixing that!

**Comment by rti** [comment]
> I think it's because I only check the first extension table. Whoops. Fixing that!

It works for others, e.g. wikibase id 2.

**Review by rti — APPROVED** [review]
With the `wikibase_software` table cleanup (if it makes sense) and the extension list parsing fix for wikibase id 1, I think this is looking good and ready to be merged.

**Pull request merged**